io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -486,6 +458,47 @@ sealed abstract class IO[E, A] { self =>
 }
 
 object IO {
+
+  trait Disj[A, B] {
+    def fold[Z](left: A => Z, right: B => Z): Z
+  }
+
+  object Disj {
+    def either[A, B](e: Either[A, B]): Disj[A, B] =
+      new Disj[A, B] {
```

Comment:
This should use a preallocated object, cast to the right type.

Reply:
I can't. The instance has to wrap the value already so no way I can use a preallocated object for this. In order to be able to do that we should add the original disjunction value as a parameter to `fold`, but that makes `Disj` API uncomfortable or unusable for other cases.

Reply:
Ah, you're right. 👍